JP Onboarding: Make saveJetpackOnboardingSettings() generic, use to perform JPC-auth'd requests

ockham opened the pull request:
Make `saveJetpackOnboardingSettings()` and `requestJetpackOnboardingSettings()` more generic in order to make them work without JPO auth tokens when connected. We now simply pass the JPO authentication token as an argument to these actions; we omit them if we're already connected to WP.com (per JPC). This means that this PR isn't simply a refactor: Instead, it allows to run the JPO flow without JPO token when connected to WP.com.

Some architectural choices:
* I was trying to re-implement `saveJetpackOnboardingSettings()` and `requestJetpackOnboardingSettings()` by just hard-wiring the token to more generic `saveJetpackSettings()` and `requestJetpackSettings() actions, respectively. However, that requires obtaining the token via `getState()`, i.e. it would cause `saveJetpackOnboardingSettings()` and `requestJetpackOnboardingSettings()` to be thunks. However, we cannot use the `getRequest()` selector (which we use to keep track of `isLoading` state) with thunks -- it only works with normal Redux action objects. Thus, I opted to just provide  `saveJetpackSettings()` and `requestJetpackSettings()` and leave it to the consumer to provide the onboarding token.
* This actually works quite nicely in the `<Main />` component's `connect()` clause, where we can centralize the logic that determines whether the site is connected and either omit or provide the onboarding token accordingly.

Note that this makes reducers and actions in `state/jetpackOnboarding` somewhat redundant with what is found in `state/jetpack/settings`. However, I opted to keep "ours" (`state/jetpackOnboarding`)  since it uses the data layer (and is arguably more elegant 😉). In a subsequent PR, I'd like to change consumers of `state/jetpack/settings` to use "ours" instead.

Also note that the directory is still called `state/jetpackOnboarding`, even though it can now generically work with JP settings.

Finally, note that `saveJetpackOnboardingSettings()` and `requestJetpackOnboardingSettings()` will be renamed to `saveJetpackSettings()` and `requestJetpackSettings()`, respectively, in the follow-up PR that will deduplicate `state/jetpack/settings` (alongside action types etc). As per discussion with @tyxla, I'm not doing that as part of this PR in order to somewhat separate PR concerns, and to keep diff size down.

To test:
This requires some thorough testing, since it touches almost all of JPO state.
* Verify that JPO (while _not connected_ to WP.com) still works.
* In particular, verify that WooCommerce install retries still work on a slow host (try Pressable) (this is to verify the `retryOrAnnounceSaveFailure()` function still works).
* Also, pay attention to the summary page, and whether checkmarks/crosses/spinners are shown correctly for each individual step.
* Verify that JPO also works while connected to WP.com.

Fixes #22578 🙂

ockham commented:
@tyxla found this issue (with clean Redux state, and `sleep(3);` added to the top of `jetpack.php`:

![_mie1weh6s](https://user-images.githubusercontent.com/96308/36797145-29776130-1c9f-11e8-926c-926ae79e1554.png)

ockham commented:
The underlying reason seems to be that the data-layer middleware changes  `requesting` state _before_ we're able to fire `updateJetpackOnboardingSettings` 🙁

ockham commented:
Still wondering why this wasn't an issue before 🤔

ockham commented:
It took a while, but I’ve finally tracked down the error. My initial hunch was about right — the issue really is due to `isJetpackSite()` returning different values at different times, thus causing JPO credentials to be passed or omitted.

The exact scenario I was able to repro is as follows:

1. We’re dealing with an already connected JP site.
1. We reset its Redux state and land at …, causing the usual redirects to JPO.
1. Main’s `isConnected = isJetpackSite()` cannot yet get any site from Redux state, so it’s set to `false`.
1. This causes `jpoAuth` to be set to credentials (instead of just `undefined`); it is, among other things, passed on to the query component, which makes a request using these credentials.
1. In `Main`’s `connect()`, `isRequestingSettings` monitors that request (i.e. the one for `requestJetpackOnboardingSettings( siteId, jpoAuth )`)
1. After some time, we finally get site data for the JP site, which flips `isConnected` to true; this causes `jpoAuth` to be unset, and `isRequestingSettings` to be set to `getRequest( state, requestJetpackOnboardingSettings( siteId, undefined ) ) )`.
1. Obviously, there is no request underway for `requestJetpackOnboardingSettings( siteId, undefined )`, so `isRequestingSettings` is flipped to `false`.
1. We now have all the ingredients to confuse `SiteTitle`’s `componentWillReceiveProps`: `isRequestingSettings` was flipped from `true` to `false`, even tho `settings` haven’t been actually received yet.

I believe the fix will be to ensure that no requests are made before we actually know whether a site is a JP site, or not.

ockham commented:
Should be finally ready for review 😅

tyxla commented:
Nice sleuthing 🕵️  😅 

Will have a good thorough testing and review today.

tyxla commented on client/state/data-layer/wpcom/jetpack/settings/index.js:
```diff
@@ -149,7 +130,7 @@ export const retryOrAnnounceSaveFailure = ( { dispatch }, action, { message: err
 	// properly, in which case a subsequent request will return 'success'.
 	if (
-		get( settings, 'installWooCommerce' ) !== true ||
+		get( settings, [ 'onboarding', 'installWooCommerce' ] ) !== true ||
```
If we keep those handlers generic, I think this is a pretty specific functionality to be defined in such a location. Could it make more sense to move it in some sort of middleware (not necessarily in this PR)?

ockham replied:
Should be done on this PR since it's where it's introduced. 👉11b3121

ockham replied:
Yep, this is more specific than anything in this file should be. I just haven't really figured out how to elegantly solve this otherwise. A middleware could be an option, I'm just not too seasoned in writing those, and I hope that middleware on top of (data-layer) middleware won't cause any problems. Anyway, warrants a separate PR IMO.

tyxla commented on client/state/data-layer/wpcom/jetpack/settings/test/index.js:
```diff
@@ -35,18 +35,7 @@ describe( 'requestJetpackOnboardingSettings()', () => {
 	};
 
 	test( 'should dispatch an action for GET HTTP request to save Jetpack Onboarding settings', () => {
```
Let's alter the test names here - these are no longer an onboarding-specific tests (this applies for most of the rest of the onboarding tests in this file).

Also, I just realized this test name has been wrong all along - we don't `save` settings - we `request` or `retrieve` them.

ockham replied:
0350947

tyxla commented on client/state/data-layer/wpcom/jetpack/settings/test/index.js:
```diff
@@ -70,14 +54,16 @@ describe( 'requestJetpackOnboardingSettings()', () => {
-		const getState = () => ( {
-			jetpackOnboarding: {
-				credentials: {},
+	test( 'should dispatch an action for GET HTTP request with a query including token and jpUser', () => {
```
Let's specify what these `token` and `jpUser` are, either by prepending them in the test name with `onboarding`, or just change them to `onboarding credentials` or `onboarding auth credentials`.

ockham replied:
Yep, excellent call. 7d7c8d8

tyxla commented on client/state/jetpack-onboarding/reducer.js:
```diff
@@ -27,12 +32,10 @@ export const settingsReducer = keyedReducer(
-				...state,
-				...settings,
-			} ),
+			[ JETPACK_ONBOARDING_SETTINGS_UPDATE ]: ( state, { settings } ) =>
```
We're now storing all of Jetpack settings, but in the onboarding settings tree? 😭

ockham replied:
Yes! Another total misnomer as of this PR. There's already `state.jetpack.settings` for the Redux stuff in, well, `state/jetpack/settings`, so we can't rename it to that just yet. And renaming to `jetpackSettings` for now didn't seem to make much sense if we're going to replace the stuff in `state/jetpack/settings` with it anyway, so I opted to keep all misnomers for now and address them in the de-dupe PR.

tyxla commented on client/jetpack-onboarding/main.jsx:
```diff
@@ -113,8 +132,27 @@ export default connect(
+			isRequestingSite( state, siteId ) || isRequestingSites( state );
+		const isConnected = isJetpackSite( state, siteId );
+
+		let jpoAuth;
```
With approach of having the logic to decide whether to pass onboarding auth credentials to the actions placed in the component, we lack tests for this logic. I think it is pretty important to have tests for this, as it's essential for allowing JPO requests to work with both connected and unconnected sites. Do you think we can add some component tests for this? Could happen in a subsequent PR, of course.

ockham replied:
Testing `connect()`ed components used to be a bit of a pain in the past (though I haven't revisited that sort of problem in a while). If we go with a separate selector as you suggest [here](https://github.com/Automattic/wp-calypso/pull/22775#discussion_r171494411), we might be able to cover that with unit tests for all these cases.

ockham commented on client/components/data/query-jetpack-onboarding-settings/index.jsx:
```diff
@@ -15,6 +15,7 @@ import { requestJetpackOnboardingSettings } from 'state/jetpack-onboarding/actio
 
 class QueryJetpackOnboardingSettings extends Component {
 	static propTypes = {
+		query: PropTypes.object,
```
Should be done on this PR since it's where it's introduced. 👉11b3121

ockham commented on client/jetpack-onboarding/main.jsx:
```diff
@@ -113,8 +132,27 @@ export default connect(
+
+		let jpoAuth;
+
+		if ( ! isConnected && getUnconnectedSite( state, siteId ) ) {
```
Good point. Would opt for another PR in this case, since it'll be a non-trivial addition involving writing unit tests.

ockham commented:
I think I've addressed all of your feedback now @tyxla -- thanks a bunch, lots of great suggestions!

I acknowledge that this does make pretty much the entire `jetpackOnboarding` subtree a misnomer, but renaming things to something that's closer to "jetpack settings" would either cause collisions or confusion with the existing `jetpack/settings` Redux tree, so I really think we best do the renaming as part of the deduplication and live with the misnomer until then.
